File: Python_Project/projectops/services/project_service.py
from urllib.parse import urlparse

CATEGORIES = ['CI/CD', 'Docker', 'Kubernetes', 'Terraform', 'Ansible', 'Cloud', 'DevSecOps', 'Monitoring', 'GitOps', 'Linux', 'Other']
DIFFICULTIES = ['Beginner', 'Intermediate', 'Advanced']
STATUSES = ['Planned', 'Active', 'Completed', 'On Hold']


def clean(value):
    return value.strip() if isinstance(value, str) else ''


def normalize_project(data=None):
    data = data or {}
    return {
        'title': clean(data.get('title')),
        'category': clean(data.get('category')),
        'difficulty': clean(data.get('difficulty')),
        'status': clean(data.get('status')),
        'owner': clean(data.get('owner')),
        'description': clean(data.get('description')),
        'repository_url': clean(data.get('repository_url')) or None,
    }


def validate_project(data=None):
    project = normalize_project(data)
    errors = {}
    if not 3 <= len(project['title']) <= 120:
        errors['title'] = 'Title must be between 3 and 120 characters.'
    if project['category'] not in CATEGORIES:
        errors['category'] = 'Choose a valid category.'
    if project['difficulty'] not in DIFFICULTIES:
        errors['difficulty'] = 'Choose a valid difficulty.'
    if project['status'] not in STATUSES:
        errors['status'] = 'Choose a valid status.'
    if not 2 <= len(project['owner']) <= 80:
        errors['owner'] = 'Owner must be between 2 and 80 characters.'
    if not 10 <= len(project['description']) <= 800:
        errors['description'] = 'Description must be between 10 and 800 characters.'
    if project['repository_url']:
        parsed = urlparse(project['repository_url'])
        if parsed.scheme not in ('http', 'https') or not parsed.netloc:
            errors['repository_url'] = 'Repository URL must be a valid http/https URL.'
    return project, errors


def row_to_dict(row):
    return dict(row) if row is not None else None


class ProjectRepository:
    def __init__(self, db):
        self.db = db

    def list(self, filters=None):
        filters = filters or {}
        clauses, values = [], []
        search = clean(filters.get('search'))
        status = clean(filters.get('status'))
        category = clean(filters.get('category'))
        difficulty = clean(filters.get('difficulty'))

        if search:
            clauses.append('(LOWER(title) LIKE LOWER(?) OR LOWER(owner) LIKE LOWER(?) OR LOWER(description) LIKE LOWER(?))')
            token = f'%{search}%'
            values.extend([token, token, token])
        if status in STATUSES:
            clauses.append('status = ?'); values.append(status)
        if category in CATEGORIES:
            clauses.append('category = ?'); values.append(category)
        if difficulty in DIFFICULTIES:
            clauses.append('difficulty = ?'); values.append(difficulty)

        where = f"WHERE {' AND '.join(clauses)}" if clauses else ''
        rows = self.db.execute(f'SELECT * FROM projects {where} ORDER BY datetime(updated_at) DESC, id DESC', values).fetchall()
        return [row_to_dict(row) for row in rows]

    def find_by_id(self, project_id):
        row = self.db.execute('SELECT * FROM projects WHERE id = ?', (int(project_id),)).fetchone()
        return row_to_dict(row)

    def create(self, data):
        project, errors = validate_project(data)
        if errors:
            return None, errors
        cursor = self.db.execute('''
            INSERT INTO projects (title, category, difficulty, status, owner, description, repository_url)
            VALUES (?, ?, ?, ?, ?, ?, ?)
        ''', tuple(project.values()))
        self.db.commit()
        return self.find_by_id(cursor.lastrowid), None

    def update(self, project_id, data):
        if not self.find_by_id(project_id):
            return None, None, True
        project, errors = validate_project(data)
        if errors:
            return None, errors, False
        self.db.execute('''
            UPDATE projects
            SET title=?, category=?, difficulty=?, status=?, owner=?, description=?, repository_url=?, updated_at=datetime('now')
            WHERE id=?
        ''', (*project.values(), int(project_id)))
        self.db.commit()
        return self.find_by_id(project_id), None, False

    def remove(self, project_id):
        if not self.find_by_id(project_id):
            return False
        self.db.execute('DELETE FROM projects WHERE id = ?', (int(project_id),))
        self.db.commit()
        return True

    def stats(self):
        scalar = lambda sql: int(self.db.execute(sql).fetchone()[0])
        by_status = [dict(row) for row in self.db.execute('SELECT status, COUNT(*) AS count FROM projects GROUP BY status ORDER BY count DESC')]
        by_category = [dict(row) for row in self.db.execute('SELECT category, COUNT(*) AS count FROM projects GROUP BY category ORDER BY count DESC, category')]
        return {
            'total': scalar('SELECT COUNT(*) FROM projects'),
            'active': scalar("SELECT COUNT(*) FROM projects WHERE status='Active'"),
            'completed': scalar("SELECT COUNT(*) FROM projects WHERE status='Completed'"),
            'advanced': scalar("SELECT COUNT(*) FROM projects WHERE difficulty='Advanced'"),
            'categories': scalar('SELECT COUNT(DISTINCT category) FROM projects'),
            'byStatus': by_status,
            'byCategory': by_category,
        }

File: NodeJS_Project/src/services/projectService.js
const CATEGORIES = ['CI/CD', 'Docker', 'Kubernetes', 'Terraform', 'Ansible', 'Cloud', 'DevSecOps', 'Monitoring', 'GitOps', 'Linux', 'Other'];
const DIFFICULTIES = ['Beginner', 'Intermediate', 'Advanced'];
const STATUSES = ['Planned', 'Active', 'Completed', 'On Hold'];

function clean(value) {
  return typeof value === 'string' ? value.trim() : '';
}

function normalizeProject(input = {}) {
  return {
    title: clean(input.title),
    category: clean(input.category),
    difficulty: clean(input.difficulty),
    status: clean(input.status),
    owner: clean(input.owner),
    description: clean(input.description),
    repository_url: clean(input.repository_url) || null
  };
}

function validateProject(input = {}) {
  const project = normalizeProject(input);
  const errors = {};

  if (project.title.length < 3 || project.title.length > 120) {
    errors.title = 'Title must be between 3 and 120 characters.';
  }
  if (!CATEGORIES.includes(project.category)) {
    errors.category = 'Choose a valid category.';
  }
  if (!DIFFICULTIES.includes(project.difficulty)) {
    errors.difficulty = 'Choose a valid difficulty.';
  }
  if (!STATUSES.includes(project.status)) {
    errors.status = 'Choose a valid status.';
  }
  if (project.owner.length < 2 || project.owner.length > 80) {
    errors.owner = 'Owner must be between 2 and 80 characters.';
  }
  if (project.description.length < 10 || project.description.length > 800) {
    errors.description = 'Description must be between 10 and 800 characters.';
  }
  if (project.repository_url) {
    try {
      const url = new URL(project.repository_url);
      if (!['http:', 'https:'].includes(url.protocol)) throw new Error('Invalid protocol');
    } catch {
      errors.repository_url = 'Repository URL must be a valid http/https URL.';
    }
  }

  return { valid: Object.keys(errors).length === 0, errors, project };
}

function buildProjectRepository(db) {
  return {
    list(filters = {}) {
      const clauses = [];
      const values = [];
      const search = clean(filters.search);
      const status = clean(filters.status);
      const category = clean(filters.category);
      const difficulty = clean(filters.difficulty);

      if (search) {
        clauses.push('(LOWER(title) LIKE LOWER(?) OR LOWER(owner) LIKE LOWER(?) OR LOWER(description) LIKE LOWER(?))');
        const token = `%${search}%`;
        values.push(token, token, token);
      }
      if (status && STATUSES.includes(status)) {
        clauses.push('status = ?');
        values.push(status);
      }
      if (category && CATEGORIES.includes(category)) {
        clauses.push('category = ?');
        values.push(category);
      }
      if (difficulty && DIFFICULTIES.includes(difficulty)) {
        clauses.push('difficulty = ?');
        values.push(difficulty);
      }

      const where = clauses.length ? `WHERE ${clauses.join(' AND ')}` : '';
      return db.prepare(`SELECT * FROM projects ${where} ORDER BY datetime(updated_at) DESC, id DESC`).all(...values);
    },

    findById(id) {
      return db.prepare('SELECT * FROM projects WHERE id = ?').get(Number(id));
    },

    create(input) {
      const { valid, errors, project } = validateProject(input);
      if (!valid) return { errors };

      const result = db.prepare(`
        INSERT INTO projects
          (title, category, difficulty, status, owner, description, repository_url)
        VALUES (?, ?, ?, ?, ?, ?, ?)
      `).run(
        project.title,
        project.category,
        project.difficulty,
        project.status,
        project.owner,
        project.description,
        project.repository_url
      );
      return { project: this.findById(result.lastInsertRowid) };
    },

    update(id, input) {
      const existing = this.findById(id);
      if (!existing) return { notFound: true };

      const { valid, errors, project } = validateProject(input);
      if (!valid) return { errors };

      db.prepare(`
        UPDATE projects
        SET title = ?, category = ?, difficulty = ?, status = ?, owner = ?,
            description = ?, repository_url = ?, updated_at = datetime('now')
        WHERE id = ?
      `).run(
        project.title,
        project.category,
        project.difficulty,
        project.status,
        project.owner,
        project.description,
        project.repository_url,
        Number(id)
      );
      return { project: this.findById(id) };
    },

    remove(id) {
      const existing = this.findById(id);
      if (!existing) return false;
      db.prepare('DELETE FROM projects WHERE id = ?').run(Number(id));
      return true;
    },

    stats() {
      const total = db.prepare('SELECT COUNT(*) AS count FROM projects').get().count;
      const active = db.prepare("SELECT COUNT(*) AS count FROM projects WHERE status = 'Active'").get().count;
      const completed = db.prepare("SELECT COUNT(*) AS count FROM projects WHERE status = 'Completed'").get().count;
      const advanced = db.prepare("SELECT COUNT(*) AS count FROM projects WHERE difficulty = 'Advanced'").get().count;
      const categories = db.prepare('SELECT COUNT(DISTINCT category) AS count FROM projects').get().count;
      const byStatus = db.prepare('SELECT status, COUNT(*) AS count FROM projects GROUP BY status ORDER BY count DESC').all();
      const byCategory = db.prepare('SELECT category, COUNT(*) AS count FROM projects GROUP BY category ORDER BY count DESC, category').all();

      return {
        total: Number(total),
        active: Number(active),
        completed: Number(completed),
        advanced: Number(advanced),
        categories: Number(categories),
        byStatus,
        byCategory
      };
    }
  };
}

module.exports = {
  CATEGORIES,
  DIFFICULTIES,
  STATUSES,
  normalizeProject,
  validateProject,
  buildProjectRepository
};
